snapcraft/internal/project_loader/_extensions/gnome-3-28.py

```diff
@@ -0,0 +1,111 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Import types and tell flake8 to ignore the "unused" List.
+from typing import Any, Dict, List  # noqa: F401
+
+from snapcraft.internal import errors
+from ._desktop_common import DesktopCommonExtension
+
+
+class GnomeExtensionClassicConfinementError(errors.SnapcraftError):
+    fmt = "The gnome extension doesn't support classic confinement."
+
+
+class Gnome_3_28Extension(DesktopCommonExtension):
+    """The Gnome extension.
+    This extension is to be used by applications that require GTK+.
+    Examples might include productivity applications or utilities.
+    Note that this extension does not support classically-confined snaps at this time.
+    """
+
+    supported_bases = ("core18",)
+
+    def __init__(self, yaml_data: Dict[str, Any]) -> None:
+        """Create a new GnomeExtension.
+        Note that this extension does not support classic snaps.
+        :param dict yaml_data: Loaded snapcraft.yaml data.
+        """
+
+        super().__init__(yaml_data)
+
+        if yaml_data.get("confinement") == "classic":
+            raise GnomeExtensionClassicConfinementError()
+
+        platform_snap = ""
+        base = yaml_data.get("base")
+        after_dependencies = []  # type: List[str]
+        dependency_part = {}  # type: Dict[str, Any]
+        if base is not None:
+            if base == "core18":
+                platform_snap = "gnome-3-28-1804"
+
+        layout = {
+            "layout": {
+                "/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/webkit2gtk-4.0": {
+                    "bind": "$SNAP/gnome-platform/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/webkit2gtk-4.0"
+                },
+                "/usr/share/xml/iso-codes": {
+                    "symlink": "$SNAP/gnome-platform/usr/share/xml/iso-codes"
+                },
+            }
+        }  # type: Dict[str, Any]
+
+        self.root_snippet = {
+            **self.root_snippet,
+            **layout,
+            "plugs": {
+                **self.plugs,
+                platform_snap: {
+                    "interface": "content",
+                    "target": "$SNAP/gnome-platform",
+                    "default-provider": "{snap}:{slot}".format(
+                        snap=platform_snap, slot=platform_snap
+                    ),
+                },
+            },
+            "environment": {
+                **self.environment,
+                "SNAP_DESKTOP_RUNTIME": "$SNAP/gnome-platform",
+            },
+        }
+
+        command_chain = self.app_snippet["command-chain"]
+        exec_command = command_chain.pop()
```

Comment:
This logic seems fragile. It assumes the implementation of a complete separate piece which can change somewhat easily. We might drop the command from the base, and put it here by name explicitly instead.

Reply:
I like your idea. I felt dirty when I wrote that :-p

Reply:
sorted this out in ab206008. Now the last item which I was popping off the list is gone entirely in favour of duplicating the three lines in one script for each of the desktop extensions. (If we follow my current pattern of a single `desktop-gnome-specific` script then the three lines go right before the `exec`.